go-controller/cmd/ovnkube-trace/ovnkube-trace.go

```diff
@@ -0,0 +1,970 @@
+	StorMAC                 string
+	HostNetwork             bool
+}
+
+func execInPod(coreclient *corev1client.CoreV1Client, restconfig *rest.Config, namespace string, podName string, containerName string, cmd string, in string) (string, string, error) {
+
+	scheme := runtime.NewScheme()
+	if err := kapi.AddToScheme(scheme); err != nil {
+		fmt.Printf("error adding to scheme: %v", err)
+		os.Exit(-1)
+	}
+	parameterCodec := runtime.NewParameterCodec(scheme)
+
+	useStdin := false
+	if in != "" {
+		useStdin = true
+	}
+
+	// Prepare the API URL used to execute another process within the Pod.
+	req := coreclient.RESTClient().
+		Post().
+		Namespace(namespace).
+		Resource("pods").
+		Name(podName).
+		SubResource("exec").
+		VersionedParams(&kapi.PodExecOptions{
+			Container: containerName,
+			Command:   []string{"bash", "-c", cmd},
+			Stdin:     useStdin,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, parameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(restconfig, "POST", req.URL())
+	if err != nil {
+		return "", "", err
+	}
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	var stdin io.Reader
+
+	if useStdin {
+		stdin = strings.NewReader(in)
+	} else {
+		stdin = nil
+	}
+
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  stdin,
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+	if err != nil {
+		return stdout.String(), stderr.String(), err
+	}
+
+	return stdout.String(), stderr.String(), err
+}
+
+func getPodMAC(client *corev1client.CoreV1Client, pod *kapi.Pod) (podMAC string, err error) {
+
+	if pod.Spec.HostNetwork {
+		node, err := client.Nodes().Get(context.TODO(), pod.Spec.NodeName, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+
+		nodeMAC, err := util.ParseNodeManagementPortMACAddress(node)
+		if err != nil {
+			return "", err
+		}
+		if nodeMAC != nil {
+			podMAC = nodeMAC.String()
+		}
+	} else {
+		podAnnotation, err := util.UnmarshalPodAnnotation(pod.ObjectMeta.Annotations)
+		if err != nil {
+			return "", err
+		}
+		if podAnnotation != nil {
+			podMAC = podAnnotation.MAC.String()
+		}
+	}
+
+	return podMAC, nil
+}
+
+func getSvcInfo(coreclient *corev1client.CoreV1Client, restconfig *rest.Config, svcName string, ovnNamespace string, namespace string, cmd string) (svcInfo *SvcInfo, err error) {
+
+	// Get service with the name supplied by svcName
+	svc, err := coreclient.Services(namespace).Get(context.TODO(), svcName, metav1.GetOptions{})
+	if err != nil {
+		klog.V(1).Infof("Service %s in namespace %s not found\n", svcName, namespace)
+		return nil, err
+	}
+	klog.V(5).Infof("==>Got service %s in namespace %s\n", svcName, namespace)
+
+	clusterIP := svc.Spec.ClusterIP
+	klog.V(5).Infof("==>Got service %s ClusterIP is %s\n", svcName, clusterIP)
+
+	svcInfo = &SvcInfo{
+		IP: svc.Spec.ClusterIP,
+	}
+
+	fieldSelector := "metadata.name=" + svcName
+	eps, err := coreclient.Endpoints(namespace).List(context.TODO(), metav1.ListOptions{FieldSelector: fieldSelector})
```

Comment:
you don't need to list or a field selector, the endpoint will have the same name than the service, and there will be only 1 endpoint per service ... EndpointSlices is different
```
eps, err := coreclient.Endpoints(namespace).Get(serviceName)
```
_check the right syntax, the point is that a Get is enough_